### ISSUE TITLE
How to train offline agent on the huge dataset (50 Million) ?

### ISSUE THREAD

**Issue by LQNew** [google-research/batch_rl#10]
Hi, I have read your paper which was published on ICML 2020, now I try to do some research on the offline image data. I have noticed that when training the online agent, such as DQN, replay buffer capacity is usually set to be 1 million, when the size of collected data is above 1 million, the new data will cover the oldest data in the replay buffer. But when training DQN on offline data, such as your contributed data, the size of data is 50M, how do I train the agent on this so huge dataset? Since memory of the machine is limited,  it's impossible to load 50M data into the memory once. I wonder that how you solved this problem and if you implement your idea in this project, please refer it to me. At last, I really appreciate your great job and your open-source code!

**Comment by agarwl** [comment]
That's correct, it's not possible to load the entire dataset into memory. So, the way the dataset is stored is in 50 files of size 1M each corresponding to replay buffers for every 4 iterations seen during training. 

Now, to load this huge dataset, I simply created a new  [FixedReplayBuffer](https://github.com/google-research/batch_rl/blob/f9b89f43add68b603ee562ebcf9dffcb6bf6f841/batch_rl/fixed_replay/replay_memory/fixed_replay_buffer.py#L35) class for loading the offline dataset which is created in the offline agents (e.g., [here](https://github.com/google-research/batch_rl/blob/master/batch_rl/fixed_replay/agents/dqn_agent.py#L83))

 which loads [5 random replay buffers of size 1M at each iteration](https://github.com/google-research/batch_rl/blob/master/batch_rl/fixed_replay/run_experiment.py#L80) from the set of 50 buffers.

That said, experiment with the entire set of 5M dataset still has a large ram requirement, so there are two possible alternatives:

1. If you want to use the entire 50M dataset, you can use the Atari dataset released by DeepMind which can be directly loaded into tensorflow without suffering from RAM issues. See this [colab](https://github.com/deepmind/deepmind-research/blob/master/rl_unplugged/atari_dqn.ipynb) for an example. 

2. If you want to stick to dopamine, I suggestion I have is to use the smaller subsampled datasets as done in Section 6 of the paper as well as a couple of other papers (e.g, this [NeurIPS'20 paper](https://arxiv.org/pdf/2006.04779.pdf), [ICLR'20 submission](https://openreview.net/forum?id=O9bnihsFfXU)).

The way to do this is very simple: simply set the size of the replay buffer to be smaller than 1M (let's say 50000, so it'll only load the first 50000 data points from the dataset files, this can done through changing the size in the `gin` file [here](https://github.com/google-research/batch_rl/blob/master/batch_rl/fixed_replay/configs/dqn.gin#L36).   


The following functions for the `FixedReplayBuffer` class would be helpful in that regards (note the clipping done to free unused RAM if replay_capacity is smaller than 1M):


```
  def _load_buffer(self, suffix):
    """Loads a OutOfGraphReplayBuffer replay buffer."""
    try:
      # pytype: disable=attribute-error
      tf.logging.info(
          f'Starting to load from ckpt {suffix} from {self._data_dir}')
      replay_buffer = circular_replay_buffer.OutOfGraphReplayBuffer(
          *self._args, **self._kwargs)
      replay_buffer.load(self._data_dir, suffix)
      # pylint: disable = protected-access
      replay_capacity = replay_buffer._replay_capacity
      tf.logging.info(f'Capacity: {replay_buffer._replay_capacity}')
      for name, array in replay_buffer._store.items():
        # This frees unused RAM if replay_capacity is smaller than 1M
        replay_buffer._store[name] = array[:replay_capacity + 100].copy()
        tf.logging.info(f'{name}: {array.shape}')
      tf.logging.info('Loaded replay buffer ckpt {} from {}'.format(
          suffix, self._data_dir))
      # pylint: enable=protected-access
      # pytype: enable=attribute-error
      return replay_buffer
    except tf.errors.NotFoundError:
      return None

  def _start_load_replay_buffers(self, num_buffers=None):
    """Submit future requests for loading buffers."""
    ckpts = gfile.ListDirectory(self._data_dir)  # pytype: disable=attribute-error
    # Assumes that the checkpoints are saved in a format CKPT_NAME.{SUFFIX}.gz
    ckpt_counters = collections.Counter(
        [name.split('.')[-2] for name in ckpts])
    # Should contain the files for add_count, action, observation, reward,
    # terminal and invalid_range
    ckpt_suffixes = [x for x in ckpt_counters if ckpt_counters[x] in [6, 7]]
    if num_buffers is not None:
      ckpt_suffixes = np.random.choice(
          ckpt_suffixes, num_buffers, replace=False)
    # Load the replay buffers in parallel
    with futures.ThreadPoolExecutor(
        max_workers=num_buffers) as thread_pool_executor:
      self._futures.extend([thread_pool_executor.submit(
          self._load_buffer, suffix) for suffix in ckpt_suffixes])

  def _load_replay_buffers(self, num_buffers=None):
    """Loads multiple checkpoints into a list of replay buffers."""
    if not self._loaded_buffers:  # pytype: disable=attribute-error
      self._replay_buffers = []
      if len(self._futures) < num_buffers:
        self._start_load_replay_buffers(num_buffers)
      for f in self._futures:
        replay_buffer = f.result()
        if replay_buffer is not None:
          self._replay_buffers.append(replay_buffer)
          self.add_count = max(replay_buffer.add_count, self.add_count)
      self._futures = []
      self._num_replay_buffers = len(self._replay_buffers)
      if self._num_replay_buffers:
        self._loaded_buffers = True
```

Please let me know if you have any more questions!

**Comment by LQNew** [comment]
> That's correct, it's not possible to load the entire dataset into memory. So, the way the dataset is stored is in 50 files of size 1M each corresponding to replay buffers for every 4 iterations seen during training.
> 
> Now, to load this huge dataset, I simply created a new [FixedReplayBuffer](https://github.com/google-research/batch_rl/blob/f9b89f43add68b603ee562ebcf9dffcb6bf6f841/batch_rl/fixed_replay/replay_memory/fixed_replay_buffer.py#L35) class for loading the offline dataset which is created in the offline agents (e.g., [here](https://github.com/google-research/batch_rl/blob/master/batch_rl/fixed_replay/agents/dqn_agent.py#L83)) which loads [5 random replay buffers of size 1M at each iteration](https://github.com/google-research/batch_rl/blob/master/batch_rl/fixed_replay/run_experiment.py#L80) from the set of 50 buffers.
> 
> That said, experiment with the loading buffers of 5M still has a large RAM requirement, so there are two possible alternatives:
> 
> 1. If you want to use the entire 50M dataset, you can use the same Atari dataset released by DeepMind which can be directly loaded into tensorflow without suffering from RAM issues. See this [colab](https://github.com/deepmind/deepmind-research/blob/master/rl_unplugged/atari_dqn.ipynb) for an example.
> 2. If you want to stick to dopamine, I suggestion I have is to use the smaller subsampled datasets as done in Section 6 of the paper as well as a couple of other papers (e.g, this [NeurIPS'20 paper](https://arxiv.org/pdf/2006.04779.pdf), [ICLR'20 submission](https://openreview.net/forum?id=O9bnihsFfXU)).
> 
> The way to do this is very simple: simply set the size of the replay buffer to be smaller than 1M (let's say 50000, so it'll only load the first 50000 data points from the dataset files, this can done through changing the size in the `gin` file [here](https://github.com/google-research/batch_rl/blob/master/batch_rl/fixed_replay/configs/dqn.gin#L36).
> 
> The following functions for the `FixedReplayBuffer` class would be helpful in that regards (note the clipping done to free unused RAM if replay_capacity is smaller than 1M):
> 
> ```
>   def _load_buffer(self, suffix):
>     """Loads a OutOfGraphReplayBuffer replay buffer."""
>     try:
>       # pytype: disable=attribute-error
>       tf.logging.info(
>           f'Starting to load from ckpt {suffix} from {self._data_dir}')
>       replay_buffer = circular_replay_buffer.OutOfGraphReplayBuffer(
>           *self._args, **self._kwargs)
>       replay_buffer.load(self._data_dir, suffix)
>       # pylint: disable = protected-access
>       replay_capacity = replay_buffer._replay_capacity
>       tf.logging.info(f'Capacity: {replay_buffer._replay_capacity}')
>       for name, array in replay_buffer._store.items():
>         # This frees unused RAM if replay_capacity is smaller than 1M
>         replay_buffer._store[name] = array[:replay_capacity + 100].copy()
>         tf.logging.info(f'{name}: {array.shape}')
>       tf.logging.info('Loaded replay buffer ckpt {} from {}'.format(
>           suffix, self._data_dir))
>       # pylint: enable=protected-access
>       # pytype: enable=attribute-error
>       return replay_buffer
>     except tf.errors.NotFoundError:
>       return None
> ```
> 
> Please let me know if you have any more questions!

Hi, agarwl, thank you for your detailed reply! I have learned a lot from your reply, now my question has been answered.

**Comment by agarwl** [comment]
Reopening this for visibility :)
